perf(system): improve commonFileName, commonFileExt, fileType, directoryPath performance

makepanic opened the pull request:
fixes #822

makepanic commented:
This changes:
* order of definition initialization in the faker construction. This allows all fake methods to use the definitions in their constructors
* mime types and file types are now precomputed to an array to allow easy usage when picking on of its items
  * **Note**: initialization uses a [Set](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) which requires IE>= 11. I didn't find any browser compatibility matrix so this shouldn't really break things

I ran a quick (500 iterations) of each changed method through [tracerbench](https://www.tracerbench.com/) to see how it behaves. `Control` is the current master and `Experiment` is this change:

[artifact-2.pdf](https://github.com/Marak/faker.js/files/5141578/artifact-2.pdf)

It found that `commonFileName`, `commonFileExt`, `fileType` and `directoryPath` improved their runtime.

makepanic commented:
![20200828-120825](https://user-images.githubusercontent.com/1205444/91553148-3a5a0680-e92d-11ea-8052-fb3ae32d6c2b.png)

Seems like most of the execution time in the system methods is in `fileName`, `commonFileName` and `filePath`. I'll try to see how I can improve them more.

makepanic commented:
I made some adjustments to simplify the `fake` method implementation. This in turn improves any methods that use fake, e.g. company.companyName or company.catchPhrase

Updated tracerbench result:

[artifact-20.pdf](https://github.com/Marak/faker.js/files/5142141/artifact-20.pdf)

Marak commented:
Thank you for your contributions @makepanic 

We really appreciate you taking the time to track this down.

There seems to be two groups of optimizations going on here. It would be better if you could modify this PR to not make any changes to `Faker.fake` ( for now ). That way we can merge the changes to `System` module first.

Our thoughts are that the best optimization here might be replacing `faker.fake` to use string interpolation ( wasn't available when faker.js was created ). See: https://github.com/Marak/faker.js/pull/896

makepanic commented:
I've rebased the PR onto the current master.

I agree it makes sense to make this change as minimal as possible. I've removed everything outside the system changes.
The changes in index are required as it's using definitions which weren't defined yet on the previous versions.

Marak commented:
Thank you for your contributions @makepanic

Merged.

Marak commented on lib/system.js:
```diff
@@ -1,45 +1,85 @@
+function System(faker) {
+  var typeSet = new Set();
+  var extensionSet = new Set();
+  var mimeTypes = faker.definitions.system.mimeTypes;
```
This seems no good.

When requiring locales individually using this syntax: `const Faker = require('faker/locale/en');`, `faker.definitions` will not yet be populated.

We'll have to revert this or figure out another solution.

RE: #1148

Marak commented on lib/system.js:
```diff
@@ -1,45 +1,85 @@
+    var str = faker.random.words() + "." + faker.system.fileExt();
+    return str
+        .toLowerCase()
+        .replace(/\W/g, "_");
```
This is no good because it will also remove the `.` for the file extension.

@makepanic - Did you actually try using any of these methods or reviewing the actual test cases?

Marak commented on lib/system.js:
```diff
@@ -1,45 +1,85 @@
-    return str;
+    return str
+        .toLowerCase()
+        .replace(/\W/g, "_");
```
Same issue here. This will remove the `.` in the file extension and replace it with a `_`....
